orca-pipelinetemplate/src/main/java/com/netflix/spinnaker/orca/pipelinetemplate/v1schema/graph/transform/DefaultVariableAssignmentTransform.java

```diff
@@ -0,0 +1,60 @@
+/*
+ * Copyright 2017 Netflix, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.netflix.spinnaker.orca.pipelinetemplate.v1schema.graph.transform;
+
+import com.netflix.spinnaker.orca.pipelinetemplate.exceptions.IllegalTemplateConfigurationException;
+import com.netflix.spinnaker.orca.pipelinetemplate.v1schema.PipelineTemplateVisitor;
+import com.netflix.spinnaker.orca.pipelinetemplate.v1schema.model.PipelineTemplate;
+import com.netflix.spinnaker.orca.pipelinetemplate.v1schema.model.PipelineTemplate.Variable;
+import com.netflix.spinnaker.orca.pipelinetemplate.v1schema.model.TemplateConfiguration;
+import org.apache.commons.lang3.StringUtils;
+
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+public class DefaultVariableAssignmentTransform implements PipelineTemplateVisitor {
+
+  TemplateConfiguration templateConfiguration;
+
+  public DefaultVariableAssignmentTransform(TemplateConfiguration templateConfiguration) {
+    this.templateConfiguration = templateConfiguration;
+  }
+
+  @Override
+  public void visitPipelineTemplate(PipelineTemplate pipelineTemplate) {
+    if (pipelineTemplate.getVariables() == null || pipelineTemplate.getVariables().isEmpty()) {
+      return;
+    }
+
+    Map<String, Object> variables = templateConfiguration.getPipeline().getVariables();
+
+    pipelineTemplate.getVariables().stream()
+      .filter(variable -> !variables.containsKey(variable.getName()) && variable.getDefaultValue() != null)
+      .forEach(variable -> variables.put(variable.getName(), variable.getDefaultValue()));
```

Comment:
Took a few read throughs to grasp what this was doing. Maybe very more readable please to make.